modules/templates/vars/vars.go

```
@@ -0,0 +1,101 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package vars
+
+import (
+	"fmt"
+	"strings"
+)
+
+// ErrWrongSyntax represents a wrong syntax with a tempate
+type ErrWrongSyntax struct {
+	Template string
+}
+
+func (err ErrWrongSyntax) Error() string {
+	return fmt.Sprintf("Wrong syntax found in %s", err.Template)
+}
+
+// IsErrWrongSyntax returns true if the error is ErrWrongSyntax
+func IsErrWrongSyntax(err error) bool {
+	_, ok := err.(ErrWrongSyntax)
+	return ok
+}
+
+// ErrNoMatchedVar represents an error that no matched vars
+type ErrNoMatchedVar struct {
+	Template string
+	Var      string
+}
+
+func (err ErrNoMatchedVar) Error() string {
+	return fmt.Sprintf("No matched variable %s found for %s", err.Var, err.Template)
+}
+
+// IsErrNoMatchedVar returns true if the error is ErrNoMatchedVar
+func IsErrNoMatchedVar(err error) bool {
+	_, ok := err.(ErrNoMatchedVar)
+	return ok
+}
+
+// Expand replaces all variables like {var} to match
+func Expand(template string, match map[string]string, subs ...string) (string, error) {
+	var (
+		buf         strings.Builder
+		keyStartPos = -1
+	)
+	for i, c := range template {
+		switch {
+		case c == '{':
+			if keyStartPos > -1 {
+				return "", ErrWrongSyntax{
+					Template: template,
+				}
+			}
+			keyStartPos = i
+		case c == '}':
+			if keyStartPos == -1 {
+				return "", ErrWrongSyntax{
+					Template: template,
+				}
+			}
+			if i-keyStartPos <= 1 {
+				return "", ErrWrongSyntax{
+					Template: template,
+				}
+			}
+
+			if len(match) == 0 {
```

1. Why is that an error case? I strongly disagree that not finding the corresponding variable should return an error.
2. Even if I count that as an error case, it is an unnecessary one as that will already be tested down below with the `v, ok := match[template[keyStartPos+1:i]]` implicitly.
3. Even more so, it is implemented incorrectly as it disregards the positional arguments.